File: ScaleformUI_Csharp/Menus/UIMenu/Panels/UIMenuColorPanel.cs
using CitizenFX.Core;
using CitizenFX.Core.Native;
using ScaleformUI.Elements;

namespace ScaleformUI.Menu
{
    public enum ColorPanelType { Hair, Makeup }
    public class UIMenuColorPanel : UIMenuPanel
    {
        public string Title { get; set; }
        public ColorPanelType ColorPanelColorType { get; set; }
        public List<SColor> CustomColors { get; private set; }
        internal int _value { get; set; }
        public event ColorPanelChangedEvent OnColorPanelChange;
        public int CurrentSelection
        {
            get
            {
                //_getValue();
                return _value;
            }
            set
            {
                _value = value;
                if (CustomColors is null)
                {
                    if (value > 63)
                        _value -= 63;
                    if (value < 0)
                        _value += 63;
                }
                else
                {
                    if (value > CustomColors.Count - 1)
                        _value -= CustomColors.Count - 1;
                    if (value < 0)
                        _value += CustomColors.Count - 1;
                }
                if (ParentItem != null && ParentItem.Parent != null && ParentItem.Parent.Visible)
                {
                    int it = ParentItem.Parent.MenuItems.IndexOf(ParentItem);
                    ParentItem.Parent.SendPanelsToItemScaleform(it, true);
                    PanelChanged();
                    ParentItem.Parent.ColorPanelChange(ParentItem, this, _value);
                }
            }
        }
        public UIMenuColorPanel(string title, ColorPanelType ColorType, int startIndex = 0)
        {
            Title = title ?? "Color Panel";
            ColorPanelColorType = ColorType;
            _value = startIndex;
        }

        public UIMenuColorPanel(string title, List<SColor> colors, int startIndex = 0)
        {
            Title = title ?? "Color Panel";
            ColorPanelColorType = (ColorPanelType)2;
            _value = startIndex;
            CustomColors = colors;
        }

        internal void PanelChanged()
        {
            OnColorPanelChange?.Invoke(ParentItem, this, CurrentSelection);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/UIMenu/panels/UIMenuColorPanel.lua
UIMenuColorPanel = setmetatable({}, UIMenuColorPanel)
UIMenuColorPanel.__index = UIMenuColorPanel
UIMenuColorPanel.__call = function() return "UIMenuPanel", "UIMenuColorPanel" end

---@class UIMenuColorPanel
---@field public Title string
---@field public ColorPanelColorType number
---@field public value number
---@field public CustomColors table<SColor>
---@field public ParentItem table
---@field public SetParentItem fun(self:UIMenuStatisticsPanel, item:UIMenuItem):UIMenuItem -- required
---@field public OnColorPanelChanged function

---New
---@param title string
---@param colorType number
---@param startIndex number
function UIMenuColorPanel.New(title, colorType, startIndex, colors)
    if colors ~= nil then
        colorType = 2
    end

    local _UIMenuColorPanel = {
        Title = title or "Color Panel",
        ColorPanelColorType = colorType,
        value = startIndex or 0,
        CustomColors = colors or nil,
        ParentItem = nil, -- required
        OnColorPanelChanged = function(item, panel, newindex)
        end
    }
    return setmetatable(_UIMenuColorPanel, UIMenuColorPanel)
end

---SetParentItem
---@param Item table
function UIMenuColorPanel:SetParentItem(Item) -- required
    if not Item() == nil then
        self.ParentItem = Item
    else
        return self.ParentItem
    end
end

function UIMenuColorPanel:CurrentSelection(new_value)
    if new_value ~= nil then
        self.value = new_value
        if self.ParentItem ~= nil and self.ParentItem:SetParentMenu() ~= nil and self.ParentItem:SetParentMenu():Visible() then
            local it = IndexOf(self.ParentItem:SetParentMenu().Items, self.ParentItem)
            self.ParentItem:SetParentMenu():SendPanelsToItemScaleform(it, true)
            self.OnColorPanelChanged(self.ParentItem, self, self.value)
            self.ParentItem:SetParentMenu().OnColorPanelChanged(self.ParentItem, self, self.value)
        end
    else
        return self.value
    end
end
